File: day14/main.dart
import 'dart:async';
import 'dart:io';
import 'dart:convert';
import 'dart:math';

Future<List<String>> getLines(filename) async {
  List<String> lines = [];
  await new File(filename)
      .openRead()
      .map(utf8.decode)
      .transform(new LineSplitter())
      .forEach((line) {
    lines.add(line);
  });
  return lines;
}

Map<String, int> process(polymer, rules, steps) {
  Map<String, int> newPolymer = {};
  for (int step = 0; step < steps; step++) {
    for (final poly in polymer.keys) {
      String? insert = rules[poly];
      if (insert != null) {
        newPolymer[poly.substring(0, 1) + insert] =
            (newPolymer[poly.substring(0, 1) + insert] ?? 0) + polymer[poly]!
                as int;
        newPolymer[insert + poly.substring(1, 2)] =
            (newPolymer[insert + poly.substring(1, 2)] ?? 0) + polymer[poly]!
                as int;
      } else {
        newPolymer[poly] = 1;
      }
    }
    polymer = newPolymer;
    newPolymer = {};
  }
  return polymer;
}

void main(List<String> arguments) async {
  String filename = 'input.txt';
  if (arguments.length == 1) {
    filename = arguments[0];
  }

  int part1 = 0, steps1 = 10;
  int part2 = 0, steps2 = 40;

  List<String> lines = await getLines(filename);
  Map<String, int> polymer1 = {}, polymer2 = {};
  Map<String, String> rules = {};

  String startvalue = lines[0];
  for (int i = 0; i < startvalue.length - 1; i++) {
    polymer1[startvalue[i] + startvalue[i + 1]] =
        (polymer1[startvalue[i] + startvalue[i + 1]] ?? 0) + 1;
  }
  polymer2 = {...polymer1};
  print(polymer1);

  for (final line in lines) {
    if (line.contains('->')) {
      rules[line.split(' -> ')[0]] = line.split(' -> ')[1];
    }
  }

  polymer1 = process(polymer1, rules, steps1);
  polymer2 = process(polymer2, rules, steps2);

  Map<String, int> results1 = {}, results2 = {};

  for (final poly in polymer1.keys) {
    results1[poly.substring(0, 1)] =
        (results1[poly.substring(0, 1)] ?? 0) + polymer1[poly]!;
    results1[poly.substring(1, 2)] =
        (results1[poly.substring(1, 2)] ?? 0) + polymer1[poly]!;
  }

  for (final poly in polymer2.keys) {
    results2[poly.substring(0, 1)] =
        (results2[poly.substring(0, 1)] ?? 0) + polymer2[poly]!;
    results2[poly.substring(1, 2)] =
        (results2[poly.substring(1, 2)] ?? 0) + polymer2[poly]!;
  }

  for (final result in results1.keys) {
    results1[result] = (results1[result]! / 2).ceil();
  }
  for (final result in results2.keys) {
    results2[result] = (results2[result]! / 2).ceil();
  }

  print(polymer1);
  print(results1);
  part1 = results1.values.reduce(max) - results1.values.reduce(min);
  part2 = results2.values.reduce(max) - results2.values.reduce(min);

  print('part 1: $part1');
  print('part 2: $part2');
}

File: day14/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"sort"
	"strings"
)

func getFromMapOrDefault(poly map[string]int, key string, def int) int {
	if value, ok := poly[key]; ok {
		return value
	}
	return def
}

func tryInitMapValue(poly map[string]int, key string, value int) map[string]int {
	if _, ok := poly[key]; !ok {
		poly[key] = value
	}
	return poly
}

func pairInsertion(polymer map[string]int, rules map[string]string, steps int) map[string]int {
	newPolymer := make(map[string]int)
	for i := 0; i < steps; i++ {
		for poly, value := range polymer {
			if insert, ok := rules[poly]; ok {
				newPolymer[poly[:1]+insert] += value
				newPolymer[insert+poly[1:]] += value
			} else {
				newPolymer[poly] = 1
			}
		}
		polymer = newPolymer
		newPolymer = make(map[string]int)
	}
	return polymer
}

func getQuantity(polymer map[string]int) int {
	results := make(map[string]int)
	for poly, count := range polymer {
		tryInitMapValue(results, poly[:1], 0)
		tryInitMapValue(results, poly[1:], 0)
		results[poly[:1]] += count
		results[poly[1:]] += count
	}
	var values []int
	for _, value := range results {
		// values = append(values, value)
		values = append(values, int(math.Ceil(float64(value)/2)))
	}
	sort.Ints(values)
	return values[len(values)-1] - values[0]
}

func main() {
	var filename string = "input.txt"
	if len(os.Args) == 2 {
		filename = os.Args[1]
	}
	input, err := os.ReadFile(filename)
	if err != nil {
		panic(err)
	}

	polymer := make(map[string]int)
	rules := make(map[string]string)
	lines := strings.Trim(string(input), "\n")
	lines = strings.Replace(lines, "\n\n", "\n", -1)
	for _, line := range strings.Split(lines, "\n") {
		if !strings.Contains(line, " -> ") {
			for i := 0; i < len(line)-1; i++ {
				var poly string = line[i : i+2]
				polymer[poly] = getFromMapOrDefault(polymer, poly, 0) + 1
			}
		} else {
			pair := strings.Split(line, " -> ")[0]
			element := strings.Split(line, " -> ")[1]
			rules[pair] = element
		}
	}
	polymer = pairInsertion(polymer, rules, 10)
	fmt.Println("Part 1: ", getQuantity(polymer))
	polymer = pairInsertion(polymer, rules, 30)
	fmt.Println("Part 2: ", getQuantity(polymer))
}
